guestcluster.go

```diff
@@ -0,0 +1,163 @@
+	}
+
+	g := &GuestCluster{
+		certsSearcher: config.CertsSearcher,
+		logger:        config.Logger,
+
+		certID: config.CertID,
+	}
+
+	return g, nil
+}
+
+// NewG8sClient returns a generated clientset for the specified guest cluster.
+func (g *GuestCluster) NewG8sClient(ctx context.Context, clusterID, apiDomain string) (versioned.Interface, error) {
+	g.logger.LogCtx(ctx, "level", "debug", "message", "creating G8s client for the guest cluster")
+
+	restConfig, err := g.newRestConfig(ctx, clusterID, apiDomain)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	g8sClient, err := versioned.NewForConfig(restConfig)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	g.logger.LogCtx(ctx, "level", "debug", "message", "created G8s client for the guest cluster")
+
+	return g8sClient, nil
+}
+
+// NewHelmClient returns a Helm client for the specified guest cluster.
+func (g *GuestCluster) NewHelmClient(ctx context.Context, clusterID, apiDomain string) (helmclient.Interface, error) {
+	g.logger.LogCtx(ctx, "level", "debug", "message", "creating Helm client for the guest cluster")
+
+	restConfig, err := g.newRestConfig(ctx, clusterID, apiDomain)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	k8sClient, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	c := helmclient.Config{
+		K8sClient: k8sClient,
+		Logger:    g.logger,
+
+		RestConfig:      restConfig,
+		TillerNamespace: tillerDefaultNamespace,
+	}
+	helmClient, err := helmclient.New(c)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	g.logger.LogCtx(ctx, "level", "debug", "message", "created Helm client for the guest cluster")
+
+	return helmClient, nil
+}
+
+// NewK8sClient returns a Kubernetes clientset for the specified guest cluster.
+func (g *GuestCluster) NewK8sClient(ctx context.Context, clusterID, apiDomain string) (kubernetes.Interface, error) {
+	g.logger.LogCtx(ctx, "level", "debug", "message", "creating K8s client for the guest cluster")
+
+	restConfig, err := g.newRestConfig(ctx, clusterID, apiDomain)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	k8sClient, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	g.logger.LogCtx(ctx, "level", "debug", "message", "created K8s client for the guest cluster")
+
+	return k8sClient, nil
+}
+
+// newRestConfig returns a Kubernetes REST config for the specified guest
+// cluster.
+func (g *GuestCluster) newRestConfig(ctx context.Context, clusterID, apiDomain string) (*rest.Config, error) {
+	var err error
+
+	g.logger.LogCtx(ctx, "level", "debug", "message", "looking for certificates for the guest cluster")
+
+	var tls certs.TLS
+	{
+		tls, err = g.certsSearcher.SearchTLS(clusterID, g.certID)
+		if certs.IsTimeout(err) {
+			return nil, microerror.Maskf(timeoutError, "waited too long for certificates")
```

Comment:
All the code is basically copied from the `cluster-operator`. One thing I changed though is the error here. The interface changes so we do not accidentally miss this to fix when using the lib somewhere later on.